### ISSUE TITLE
Providing a positive user experience for slow Penrose trios

### ISSUE THREAD

**Issue by joshsunshine** [penrose/penrose#1095]
**Is your feature request related to a problem? Please describe.**
When selecting a long-running and complex Penrose trio like the `hypergraph-hypergraph` from the example tab in the IDE, the browser window hangs and shows "Loading example..." for several seconds (or more) while the trio compiles and the optimizer runs. This is not a good user experience. The user can't edit the trio while this is happening and has no sense of progress toward completion. 

**Describe the solution you'd like**
The state of the system should be constantly visible to the user. To enable that, I propose the following process:
1. Load the trio into the code window tabs before starting any compilation and optimization so at least the user can see the code of the example immediately.
2. Show a progress bar in the diagram window during compilation so the user can see that progress is being made.
3. After compilation, show the first sample in the diagram window.
4. By default, dynamically change the step size so the user can observe optimization progress as it happens. Perhaps the diagram view should change at least 5 times per second. (I chose the value 5 arbitrarily. Feel free to alter it.)

In addition, the compilation and optimization should not freeze the rest of the browser window. A user should be able to interact with the code, switch to another example, etc. during those processes. I suspect the best option to allow for this interaction is to use [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) to run compilation and optimization.

**Comment by wodeni** [comment]
Looked into web workers today and here's the status so far:
* [@shoplify/web-worker](https://www.npmjs.com/package/@shopify/react-web-worker) has a nice API, but doesn't seem to run anything on a separate thread. Moving `stepState` to a worker still blocks the main thread.
  * Probably something is wrong with the library itself: https://github.com/Shopify/quilt/issues/2232
* The native web worker API is okay, but the messages must be "clonable." Passing `PenroseState` to or from the worker will result in errors that look like:

```
Uncaught (in promise) DOMException: Failed to execute 'postMessage' on 'Worker': (rng) => randFloat(rng, min4, max4) could not be cloned.
```
* @samestep is running into the same thing for #1091. We need to sync on a possible solution.

**Comment by samestep** [comment]
Quick note: a full solution to this will probably involve getting rid of the concept of a "step size", and instead just not requiring us to (de)serialize every time we go across the Wasm boundary.

**Comment by joshsunshine** [comment]
This is the issue I brought up today @Dsm0.

**Comment by wodeni** [comment]
Closing in favor of #1254
